staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go

```diff
@@ -15,58 +15,84 @@
 package model
 
 import (
-	"k8s.io/kube-openapi/pkg/validation/spec"
 	"strings"
 
 	"gopkg.in/yaml.v3"
+	"k8s.io/apiextensions-apiserver/pkg/apiserver/schema"
 )
 
 // SchemaDeclTypes constructs a top-down set of DeclType instances whose name is derived from the root
 // type name provided on the call, if not set to a custom type.
-func SchemaDeclTypes(s *spec.Schema, maybeRootType string) (*DeclType, map[string]*DeclType) {
+func SchemaDeclTypes(s *schema.Structural, maybeRootType string) (*DeclType, map[string]*DeclType) {
 	root := SchemaDeclType(s).MaybeAssignTypeName(maybeRootType)
 	types := FieldTypeMap(maybeRootType, root)
 	return root, types
 }
 
 // SchemaDeclType returns the CEL Policy Templates type name associated with the schema element.
-func SchemaDeclType(s *spec.Schema) *DeclType {
-	if len(s.Type) < 1 {
-		return NewObjectTypeRef("*error*")
+func SchemaDeclType(s *schema.Structural) *DeclType {
+	declType, found := openAPISchemaTypes[s.Type]
+	if !found {
+		return nil
 	}
-	for _, schemaType := range s.Type {
-		declType, found := openAPISchemaTypes[schemaType]
-		if !found {
-			continue
+	if s.XIntOrString {
+		return NewObjectType("intOrString", map[string]*DeclField{
+			"strVal": {Name: "strVal", Type: StringType},
+			"intVal": {Name: "intVal", Type: IntType},
+		})
```

Comment:
how does this look like in expression later? Is there some field telling me which of the two is set?

Reply:
Example usages (from the tests):

```
has(self.strVal) && self.strVal == '100m'
has(self.intVal) && self.intVal == 1
```

Note that developers could either write this as a single compound validation rule (`... && ...`) or as two separate rules, whichever they find more convenient.

Reply:
ack, looks good